ejb3/src/main/java/org/jboss/as/ejb3/component/stateful/SessionSynchStateMachine.java

```diff
@@ -0,0 +1,98 @@
+/*
+ * JBoss, Home of Professional Open Source.
+ * Copyright 2016, Red Hat, Inc., and individual contributors
+ * as indicated by the @author tags. See the copyright.txt file in the
+ * distribution for a full listing of individual contributors.
+ *
+ * This is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of
+ * the License, or (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+
+package org.jboss.as.ejb3.component.stateful;
+
+import org.jboss.as.ejb3.logging.EjbLogger;
+
+import java.util.concurrent.atomic.AtomicStampedReference;
+
+/**
+ * Tracks the current state of an SFSB instance. Required for WFLY-6215 in order to provide serialization when a Tx reaper thread
+ * calls an SFSB's afterCompletion callback.  Simple locks are not appropriate as this could block the Tx reaper thread for the
+ * duration of a SFSB method call.
+ *
+ * @author Ryan Emerson
+ */
+class SessionSynchStateMachine {
+
+    private enum State {
+        EJB_FINISHED, // Initial State
+        EJB_RUNNING,
+        AFTER_COMPLETION_DELAYED,
+        COMPLETE // End Sate
+    }
+
+    private final AtomicStampedReference<State> stateMachine = new AtomicStampedReference<>(State.EJB_FINISHED, -1);
+    private final StatefulSessionComponentInstance instance;
+
+    SessionSynchStateMachine(StatefulSessionComponentInstance instance) {
+        this.instance = instance;
+    }
+
+    void startEJBInvocation() {
+        for (;;) {
+            State currentState = stateMachine.getReference();
```

Comment:
You're right, good catch.  There's a form of get() which accepts an int array into which the stamp can be stored (atomically).  But at this point it might be a good idea to go back to an AtomicInteger, even if it is somewhat more difficult to read, to avoid the extra array allocation and the internal extra allocations that AtomicStampedReference uses.  Use of private helper methds (like "isBitSet(mask, test)" sort of things) and well-named constants can alleviate the issue though.